vsphere/virtual_machine_config_structure.go

```diff
@@ -508,6 +509,117 @@ func flattenExtraConfig(d *schema.ResourceData, opts []types.BaseOptionValue) er
 	return d.Set("extra_config", ec)
 }
 
+// expandVAppConfig reads in all the vapp key/value pairs and returns
+// the appropriate VmConfigSpec.
+//
+// We track changes to keys to determine if any have been removed from
+// configuration - if they have, we add them with an empty value to ensure
+// they are removed from vAppConfig on the update.
+func expandVAppConfig(d *schema.ResourceData) types.BaseVmConfigSpec {
+	if !d.HasChange("vapp") {
+		return nil
+	}
+
+	var (
+		props  []types.VAppPropertySpec
+		allIds sort.StringSlice
+	)
+
+	old, new := d.GetChange("vapp")
+
+	getVAppProps := func(val interface{}) map[string]interface{} {
+		vApps := val.([]interface{})
+		if vApps != nil && len(vApps) > 0 && vApps[0] != nil {
+			vApp := vApps[0].(map[string]interface{})
+			if props, ok := vApp["properties"].(map[string]interface{}); ok {
+				return props
+			}
+		}
+		return make(map[string]interface{})
+	}
+
+	oldMap := getVAppProps(old)
+	newMap := getVAppProps(new)
+
+	for k := range oldMap {
+		allIds = append(allIds, k)
+	}
+
+	for k := range newMap {
+		if _, ok := oldMap[k]; !ok {
+			allIds = append(allIds, k)
+		}
+	}
+	sort.Sort(allIds)
+
+	for i, k := range allIds {
+		_, isOld := oldMap[k]
+		_, isNew := newMap[k]
+
+		if isNew {
```

Comment:
Annotating as I'm guessing it would be around here - just noting that we will want to somehow check to ensure the key is actually present in the vApp config before working with it. This may be here, or in another part of the function, or in another function altogether (maybe it actually is part of diff customization so that we can keep the method signature clean here and as consistent as possible with the rest of the expander/flattener logic).

Reply:
Although it changes the signature, I decided to check if the key is present in this function because it does clean up the logic of the function significantly, now that we no longer have to merge two different lists of vApp properties and can rely on the canonical list of properties from the template.